text-serializer-ansi/src/main/java/net/kyori/adventure/text/serializer/ansi/ANSIComponentSerializerImpl.java

```diff
@@ -0,0 +1,276 @@
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package net.kyori.adventure.text.serializer.ansi;
+
+import net.kyori.adventure.text.Component;
+import net.kyori.adventure.text.TextComponent;
+import net.kyori.adventure.text.TranslatableComponent;
+import net.kyori.adventure.text.format.NamedTextColor;
+import net.kyori.adventure.text.format.TextColor;
+import net.kyori.adventure.text.format.TextDecoration;
+import net.kyori.adventure.text.format.TextFormat;
+import net.kyori.adventure.text.renderer.TranslatableComponentRenderer;
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.fusesource.jansi.Ansi;
+
+import java.util.EnumSet;
+import java.util.List;
+import java.util.Locale;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * {@inheritDoc}
+ */
+final class ANSIComponentSerializerImpl implements ANSIComponentSerializer {
+  private final boolean colorDownSample;
+  static ANSIComponentSerializer TRUE_COLOR = new ANSIComponentSerializerImpl(false);
+
+  private ANSIComponentSerializerImpl(final boolean colorDownSample) {
+    this.colorDownSample = colorDownSample;
+  }
+
+  private static final TextDecoration[] DECORATIONS = TextDecoration.values();
+  private static final String RGB_FORMAT = ESC_CHAR + "[38;2;%d;%d;%dm";
+  private static final Pattern STRIP_ESC_CODES = Pattern.compile(ESC_CHAR + "\\[[;\\d]*m");
+
+  private Ansi toANSI(final TextFormat format){
+    if(format instanceof NamedTextColor || format instanceof TextDecoration){
+      return this.ansiFormat(Formats.byFormat(format));
+    }
+    if(format instanceof TextColor){
+      if(this.colorDownSample){
+        return this.ansiFormat(Formats.byFormat(NamedTextColor.nearestTo((TextColor) format)));
+      }
+      final int red = ((TextColor) format).red();
+      final int blue = ((TextColor) format).blue();
+      final int green = ((TextColor) format).green();
```

Comment:
There's a lot of casting here, add a `final TextColor color = (TextColor) format;` above this code block.